File: dev-page/src/main/java/org/example/page/task/signin/Controller.java
package org.example.page.task.signin;

import org.example.core.page.application.AbstractService;
import org.example.core.page.presentation.AbstractController;
import org.example.intent.ExpertUserSignUp;
import org.example.intent.GeneralUserSignUp;
import org.example.intent.MyPage;

import javax.swing.*;
import java.awt.*;

class Controller extends AbstractController {

    Controller(AbstractService service) {
        super(service);
    }

    @Override
    public JPanel getLayout() {
        JPanel layout = new JPanel();
        layout.setLayout(new BoxLayout(layout, BoxLayout.Y_AXIS));
        layout.setBorder(BorderFactory.createEmptyBorder(40, 60, 40, 60));

        JTextField idField = new JTextField(20);
        JPasswordField passwordField = new JPasswordField(20);

        JButton loginButton = new JButton("Login");
        JButton generalUserSignUpButton = new JButton("Sign Up as General user");
        JButton expertUserSignUpButton = new JButton("Sign Up as Expert user");

        Component[] components = {idField, passwordField, loginButton, generalUserSignUpButton, expertUserSignUpButton};
        Dimension buttonSize = new Dimension(250, 35);

        for (Component comp : components) {
            ((JComponent) comp).setAlignmentX(Component.CENTER_ALIGNMENT);
            ((JComponent) comp).setMaximumSize(buttonSize);
        }

        loginButton.addActionListener(e -> onLoginButtonClick(idField.getText(), new String(passwordField.getPassword())));
        generalUserSignUpButton.addActionListener(e -> onGeneralUserSignUpButtonClick());
        expertUserSignUpButton.addActionListener(e -> onExpertUserSignUpButtonClick());

        layout.add(new JLabel("ID"));
        layout.add(idField);
        layout.add(Box.createVerticalStrut(10));

        layout.add(new JLabel("Password"));
        layout.add(passwordField);
        layout.add(Box.createVerticalStrut(20));

        layout.add(loginButton);
        layout.add(Box.createVerticalStrut(10));
        layout.add(generalUserSignUpButton);
        layout.add(Box.createVerticalStrut(10));
        layout.add(expertUserSignUpButton);

        return layout;
    }

    private void onLoginButtonClick(String id, String pw) {
        if (service instanceof Service user) {
            user.login(id, pw);
            System.out.printf("Sign in. ID: %s, pw: %s%n", id, pw);
            dispatcher.dispatch(new MyPage());
        }
    }

    private void onGeneralUserSignUpButtonClick() {
        System.out.println("general user Sign up");
        dispatcher.dispatch(new GeneralUserSignUp());
    }

    private void onExpertUserSignUpButtonClick() {
        System.out.println("expert user Sign up");
        dispatcher.dispatch(new ExpertUserSignUp());
    }
}

File: dev-page/src/main/kotlin/org/example/page/task/signin/Controller.kt
package org.example.page.task.signin

import org.example.core.page.presentation.AbstractController
import org.example.intent.ExpertUserSignUp
import org.example.intent.GeneralUserSignUp
import org.example.intent.MyPage
import java.awt.Component
import java.awt.Dimension
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.*

internal class Controller(service: Service) : AbstractController(service) {
    override val layout: JPanel
        get() {
            println("why")
            val layout = JPanel()
            layout.setLayout(BoxLayout(layout, BoxLayout.Y_AXIS))
            layout.setBorder(BorderFactory.createEmptyBorder(40, 60, 40, 60))

            val idField = JTextField(20)
            val passwordField = JPasswordField(20)

            val loginButton = JButton("Login")
            val generalUserSignUpButton = JButton("Sign Up as General user")
            val expertUserSignUpButton = JButton("Sign Up as Expert user")

            val components =
                arrayOf<Component>(idField, passwordField, loginButton, generalUserSignUpButton, expertUserSignUpButton)
            val buttonSize = Dimension(250, 35)

            for (comp in components) {
                (comp as JComponent).setAlignmentX(Component.CENTER_ALIGNMENT)
                comp.setMaximumSize(buttonSize)
            }

            loginButton.addActionListener(ActionListener { e: ActionEvent? ->
                onLoginButtonClick(
                    idField.getText(),
                    String(passwordField.getPassword())
                )
            })
            generalUserSignUpButton.addActionListener(ActionListener { e: ActionEvent? -> onGeneralUserSignUpButtonClick() })
            expertUserSignUpButton.addActionListener(ActionListener { e: ActionEvent? -> onExpertUserSignUpButtonClick() })

            layout.add(JLabel("ID"))
            layout.add(idField)
            layout.add(Box.createVerticalStrut(10))

            layout.add(JLabel("Password"))
            layout.add(passwordField)
            layout.add(Box.createVerticalStrut(20))

            layout.add(loginButton)
            layout.add(Box.createVerticalStrut(10))
            layout.add(generalUserSignUpButton)
            layout.add(Box.createVerticalStrut(10))
            layout.add(expertUserSignUpButton)

            return layout
        }

    private fun onLoginButtonClick(id: String?, pw: String?) {
        if (service is Service) {
            (service as Service).login(id, pw)
            System.out.printf("Sign in. ID: %s, pw: %s%n", id, pw)
            dispatcher?.dispatch(MyPage("MyPage"))
        }
    }

    private fun onGeneralUserSignUpButtonClick() {
        println("general user Sign up")
        dispatcher?.dispatch(GeneralUserSignUp("GeneralUserSignUp"))
    }

    private fun onExpertUserSignUpButtonClick() {
        println("expert user Sign up")
        dispatcher?.dispatch(ExpertUserSignUp("ExpertUserSignUp"))
    }
}